stable_baselines3/common/recurrent/buffers.py

```diff
@@ -148,34 +149,32 @@ def add(self, data: RecurrentRolloutBufferData, **kwargs) -> None:  # type: igno
             self.full = True
 
     def get(  # type: ignore[override]
-        self, batch_size: Optional[int] = None
+        self,
+        batch_time: int,
+        batch_envs: int,
     ) -> Generator[RecurrentRolloutBufferSamples, None, None]:
         assert self.full, "Rollout buffer must be full before sampling from it"
-
-        # Return everything, don't create minibatches
-        if batch_size is None:
-            batch_size = self.buffer_size * self.n_envs
-
-        if batch_size % self.buffer_size != 0:
-            raise ValueError(f"Batch size must be divisible by sequence length, but {batch_size=} and len={self.buffer_size}")
-
-        indices = th.randperm(self.n_envs)
-        adjusted_batch_size = batch_size // self.buffer_size
-
-        if adjusted_batch_size >= self.n_envs:
-            yield self._get_samples(slice(None))
-
-        start_idx = 0
-        while start_idx < self.n_envs:
-            yield self._get_samples(indices[start_idx : start_idx + adjusted_batch_size])
-            start_idx += adjusted_batch_size
+        if batch_envs >= self.n_envs:
+            for time_start in range(0, self.buffer_size, batch_time):
+                yield self._get_samples(seq_inds=slice(time_start, time_start + batch_time), batch_inds=slice(None))
+
+        else:
+            env_indices = th.randperm(self.n_envs)
+            for env_start in range(0, self.n_envs, batch_envs):
+                for time_start in range(0, self.buffer_size, batch_time):
```

Comment:
Initially I got these two loops reversed, so the state of the RNN was being scrambled during training. Unfortunately none of the tests in the test suite caught this mistake -- I'm not sure how to do that, a more difficult environment?

Reply:
One way to check this would be to track `last_seq_idx` in this class (`RecurrentRolloutBuffer`) which stores the last sequence index returned by `_get_samples`. Then you can add this check at top of the `_get_samples` function:

```python
def _get_samples(self, seq_inds, batch_inds):
  assert self.last_seq_idx is not None and self.last_seq_idx == seq_inds.start
  self.last_seq_idx = seq_inds.stop
  [...]
```

Reply:
That sounds awesome, thank you for the suggestion.

I think in the end I was wrong and this doesn't matter, because `get_rollouts` is only used during training and the RNN state is always re-initialized with the first one from the batch. It must have affected training through some other way, e.g. making batches less time-correlated (which is a thing we found mattered)

So I won't apply it; but if this had definitely been a problem I would have applied it.